src/values/AbstractValue.js

```diff
@@ -316,36 +290,74 @@ export default class AbstractValue extends Value {
     if (!this.mightBeEmpty) return this;
     let cond = AbstractValue.createFromBinaryOp(this.$Realm, "===", this, this.$Realm.intrinsics.empty);
     let result = AbstractValue.createFromConditionalOp(this.$Realm, cond, this.$Realm.intrinsics.undefined, this);
-    result.values = this.values.promoteEmptyToUndefined();
+    if (result instanceof AbstractValue) result.values = this.values.promoteEmptyToUndefined();
     return result;
   }
 
   refineWithPathCondition(): Value {
-    let op = this.kind;
-    if (op === "&&" || op === "||") {
-      let [left, right] = this.args;
-      let refinedLeft = left instanceof AbstractValue ? left.refineWithPathCondition() : left;
-      let refinedRight = right instanceof AbstractValue ? right.refineWithPathCondition() : right;
-      if (left === refinedLeft && right === refinedRight) return this;
-      return AbstractValue.createFromLogicalOp(this.$Realm, op, refinedLeft, refinedRight, this.expressionLocation);
+    function pathImplies(condition: AbstractValue): boolean {
+      let path = condition.$Realm.pathConditions;
+      for (let i = path.length - 1; i >= 0; i--) {
+        let pathCondition = path[i];
+        if (pathCondition.implies(condition)) return true;
+      }
+      return false;
     }
-    if (op !== "conditional") return this;
-    let [condition, trueVal, falseVal] = this.args;
-    invariant(condition instanceof AbstractValue);
-    invariant(trueVal !== undefined);
-    invariant(falseVal !== undefined);
-    let inverseCondition = AbstractValue.createFromUnaryOp(this.$Realm, "!", condition);
-    let result = this;
-    for (let pathCondition of this.$Realm.pathConditions) {
-      if (pathCondition.implies(condition)) {
-        result = trueVal;
-        break;
+
+    let realm = this.$Realm;
+    if (realm.pathConditions.length === 0) return this;
+    let op = this.kind;
+    let result = (() => {
+      if (op === "&&" || op === "||") {
+        let [left, right] = this.args;
+        invariant(left instanceof AbstractValue); // otherwise the factory would have simplified it.
+        let refinedLeft = left.refineWithPathCondition();
+        let refinedRight = right instanceof AbstractValue ? right.refineWithPathCondition() : right;
+        // todo: remove this check when there is an alternative way to indicate that an intrinsic object is nullable
+        if (!(refinedLeft instanceof AbstractObjectValue && refinedLeft.isIntrinsic())) {
+          // true && y <=> y
+          // true || y <=> true
+          if (!refinedLeft.mightNotBeTrue()) return op === "&&" ? refinedRight : realm.intrinsics.true;
```

Comment:
Why is it `realm.intrinsics.true` not `refinedLeft`?

Reply:
It should be refinedLeft.